### ISSUE TITLE
Compilation errors when trying to use watson-developer-cloud inside standard typescript project

### ISSUE THREAD

**Issue by breathe** [watson-developer-cloud/node-sdk#614]
I created a simple sample repository to demonstrate the issues I'm running into with trying to use latest watson-developer-cloud in standard typescript project.  https://github.com/breathe/sample-watson-typescript -- instructions to reproduce in the README.

The majority of the build errors occur because:
1. the project uses more strict compiler options than watson-developer-cloud
1. watson-developer-cloud npm package distributes the .ts files alongside the .d.ts as sibling files -- this means the .d.ts are not used -- thus when my project is constructing the type definitions for watson-developer-cloud the .ts files must type-check according to _my_ project's tsconfig.json rather than according to the watson-developer-cloud's tsconfig.json settings.  If the .ts files were not present, then my project would use the generated .d.ts files to get the types defined by watson-developer-cloud and since these are pre-generated according to the rules specified in watson-developer-cloud's tsconfig.json, this issue would go away.  Some context on the compiler behaviors involved here: https://github.com/Microsoft/TypeScript/issues/10704

The other category of errors occur because the published version of watson-developer-cloud requires upstream projects to set the 'dom' lib in tsconfig.json:compilerOptions.lib in order for the the ReadableStream interface type to be defined in global scope.  watson-developer-cloud depends on this type declaration from the 'dom' library -- it doesn't use or isn't compatible with the definition of the version of this interface from the `@types/node` project.  From what I can tell the version of this type defined in `@types/node` is in the NodeJs namespace -- so maybe using NodeJs.ReadableStream everywhere in watson-developer-cloud could resolve this ...?

Needing to include the 'dom' lib in requiring project's tsconfig.json isn't ideal for node projects as  including 'dom' introduces global type definitions for browser api's into the typescript compilation context.  Including these definitions in a project that is targeting node can cause the project to accidentally use unsupported-on-node browser features without getting type-errors ...  Is there a way to workaround this ... either in client projects or in the watson-developer-cloud code ...?

**Comment by breathe** [comment]
Some sample code in above fork toward partially addressing this -- but other changes would also be needed with that approach...

Another perhaps simpler approach might be to just remove all the .ts files from the published npm package via an npm prepublish task or via a .npmignore -- although I believe that would break source-maps in consuming projects.  I think that correct source-map behavior could be preserved with this approach by enabling inlineSourceMap in tsconfig.json rather than sourceMap.

**Comment by breathe** [comment]
I disagree with the 'support' label.

**Comment by breathe** [comment]
Does anyone have trouble replicating this issue as described using the sample project ...?  As is 3.x branch of this sdk is not useable from most typescript projects -- I feel like addressing this should be a priority ... ?

**Comment by anweshan** [comment]
Thanks for the input, we have been looking into this issue, apologies for the time taken to fix this.

**Comment by germanattanasio** [comment]
@breathe We moved to Typescript as part of the major release (from 2.x to 3.x). The team is not very familiar with it so this is taking us some time because we want to find the right solution. 

What do you think we should do here? I've read 2 or 3 options from your comments above. Removing the `.ts` files... updating the `tsconfig.json` file...

**Comment by breathe** [comment]
Hi @germanattanasio -- I started using typescript in mid 2017 and have converted a few projects to it and would be happy to share my thoughts.

I think maybe its two separate issues here with separate fixes:

## 1. ensure distributed npm bundle uses .d.ts rather than .ts for exported module types

A few possible solutions I can see for this (I'm not sure of all the dependencies that might exist on other artifacts/tooling outside the repo which might contribute to making some of these harder/easier to implement than others ...)

1. Option 1 (least invasive) - remove the .ts files from the npm package (via .npmignore or an npm publishOnly run-script)

1. Option 2 (more invasive) - configure the compiler to output the generated files to a different directory than the src files (using tsconfig.json outDir option).  This minimal version of this (as I tried to do here: https://github.com/breathe/node-sdk/commit/fab7b64b2a9ee89278c892993e5f14e3ee706210) will run run into issues with require path's like this one from requestwrapper.ts

```
const pkg = require('../package.json');
```

The pattern I've seen many projects adopt is to create a top-level src/ directory and move all the source files into it, then set outDir to be a sibling directory named 'lib' or  'compiled' or 'output' or something similar.  This way relative paths will work the same way from the src and the generated directories which keeps things simple conceptually.  It also has the nice effect of moving the generated products into a separate directory where they can be more easily ignored or removed via simple command -- this is especially useful if you intend to keep some of the src/ files as .js instead of .ts over the long term -- that way you can configure the typescript compiler to process those files (via the allowJs option in tsconfig) and typescript will copy the referenced .js files from src/ to the output/ directories along with the compiled ts files.  This keeps it easy to always tell at a glance if a given .js file is source of truth or generated form another source.  It also gives a nice way to reliably return to a clean directory with simple rm -rf ./output/.

1. option 3 (invasive) -- use strict typescript compiler settings in watson-developer-cloud -- this would make its unlikely that someone using the project would have stricter settings than you and find type errors in your project code during import that you did not (someone with looser compiler settings won't notice either way).  Its not perfect for this but I think this is actually reasonably workable in practice -- but it would require updating type annotations where necessary throughout the whole project to satisfy the stricter checks ...  

Regardless of utility with regard to this issue -- as you are newly adopting typescript -- I'll share my own recommendation to have a goal to enable at least the 'strict: true' option if you can.  strict: true turns on the current 'most recommended strict' compiler settings (as determined by the compiler developers) and also opts the project into future 'blessed strict' options that may be released in future typescript versions.  If that's too much type checking or too much risk when compiling with future compilers, you can selectively enable the options currently covered by the 'strict': true.  I'd aim for setting "noImplicitAny": true as a first step.  I personally find that typescript is _far_ harder to learn/use effectively with implicit any inference enabled ...

## 2. Distributed project needs to include 'dom' even when targeting node

I'm not super sure what exactly in watson-developer-cloud project is dependent on the 'dom' type declarations.  If I remove the option I get errors due to missing type definitions for 'ReadableStream' 'File', and 'Blob'.  I know 'ReadableStream' is defined on the node platorm but not sure if the type declaration from @types/node is compatible with the way the this type is used currently in watson-developer-cloud...  The other two type's don't exist on node -- I'm guessing that this code path is browser only or part of an abstraction layer that works on node and browser?

I think something along the lines of one of these two solutions is possible ... :

(1) copy type declarations for ReadableStream, File, Blob into the project (or author versions which describe the way these types are used in the project) and remove the 'dom' library from tsconfig.json.

(2) keep 'dom' in tsconfig.json but make sure that no publicly exported types use any of those types in their declaration -- this _should_ keep the dom types out of the generated .d.ts files (I think).  Then if problem 1 above is fixed such that the .d.ts files are consumed by third parties, the generated .d.ts files shouldn't depend on the types from the dom library even though internally the implementation might and so the consuming app's won't have to including the dom lib.  (I'm not 100% sure this strategy would work but I think it does ...)

**Comment by anweshan** [comment]
@breathe Thank you so much for your detailed help! I took your advice and decided to make it so the '.ts' files are no longer distributed, and copied the type declarations for ReadableStream, etc. in this pull request, and believe that it fixes compilation errors in your and other typescript projects. I did keep the 'dom' library because I think that was necessary, but I don't think it causes any errors now. https://github.com/watson-developer-cloud/node-sdk/pull/619

**Comment by breathe** [comment]
@anweshan -- awesome thanks for this!  I will give it a try soon and report back results

**Comment by anweshan** [comment]
@breathe you can clone the PR and type in `npm pack`, then unzip that into your node modules for your own project to test it out

**Comment by germanattanasio** [comment]
@breathe I just released `3.0.6` as beta.

You can install it using
```sh
npm install watson-developer-cloud@beta
```
or
```sh
npm install watson-developer-cloud@3.0.6
```

If you confirm that the fix works for you I will move the package to `latest` and close this issue

**Comment by breathe** [comment]
Ok great — sorry for delay on my part — I should be able to get to testing this later today.

**Comment by breathe** [comment]
@anweshan @germanattanasio I upgraded to the beta release and its working well for me so far.

My project's tsconfig.json:

```
{
  "compileOnSave": true,
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./lib",
    "lib": ["es2017", "es2017.object", "esnext.asynciterable"],
    "strict": true,
    "strictNullChecks": true,
    "noImplicitAny": true,
    "noImplicitThis": true,
    "alwaysStrict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "module": "commonjs",
    "target": "es2015",
    "allowJs": true,
    "experimentalDecorators": true,
    "inlineSourceMap": true,
    "inlineSources": true,
    "forceConsistentCasingInFileNames": true,
    "strictFunctionTypes": false,
    "noErrorTruncation": true
  },
  "include": ["./src/**/*"],
  "exclude": ["node_modules"]
}
```

Thank you!

**Comment by germanattanasio** [comment]
I released 3.0.6 as `latest`
